### PR TITLE
[libc][NFC] Simplify `FPBits`

### PR DESCRIPTION
**Pull request by gchatelet**
This patch reduces the scope of `FPBits` exported variables and functions.
It also moves storage up into `FPRep` and tries to make the default and specialized versions of `FPBits` more uniform.

The next step is to move the specialization from `FPBits` to `FPRep` so we can
manipulate floating point representations through `FPType` alone - that is -
independently from the host architecture.

### CODE REVIEW

**Comment by gchatelet** [comment]
I'm pushing all of the commits so it's easier to review.

File: libc/src/__support/FPUtil/FPBits.h [hunk 1]
```diff
@@ -347,17 +334,18 @@ template <typename T> struct FPBits : public internal::FPRep<get_fp_type<T>()> {
...
 
-  LIBC_INLINE constexpr FPBits() : UP() {}
+  template <typename XType> LIBC_INLINE constexpr explicit FPBits(XType x) {
+    using Unqual = typename cpp::remove_cv_t<XType>;
+    if constexpr (cpp::is_same_v<Unqual, T>) {
+      bits = cpp::bit_cast<StorageType>(x);
+    } else if constexpr (cpp::is_same_v<Unqual, StorageType>) {
+      bits = x;
+    } else {
+      static_assert(cpp::always_false<XType>);
```
Comment (legrosbuffle) [inline libc/src/__support/FPUtil/FPBits.h:327]: You probably want to keep the comment here.

File: libc/src/__support/FPUtil/x86_64/LongDoubleBits.h [hunk 2]
```diff
@@ -37,20 +37,21 @@ struct FPBits<long double> : public internal::FPRep<FPType::X86_Binary80> {
   using UP::QUIET_NAN_MASK;
 
 public:
-  static constexpr int MAX_BIASED_EXPONENT = 0x7FFF;
+  static constexpr int MAX_BIASED_EXPONENT = (1 << EXP_LEN) - 1;
+  // The x86 80 bit float represents the leading digit of the mantissa
+  // explicitly. This is the mask for that bit.
+  static constexpr StorageType EXPLICIT_BIT_MASK = StorageType(1)
+                                                   << FRACTION_LEN;
+  static_assert((EXPLICIT_BIT_MASK & FRACTION_MASK) == 0, "mask disjoint");
```
Comment (legrosbuffle) [inline libc/src/__support/FPUtil/x86_64/LongDoubleBits.h:46]:
```
// The X80 significand is made of an explicit bit and the fractional part.
static_assert((EXPLICIT_BIT_MASK & FRACTION_MASK) == 0, "the explicit bit and the fractional part should not overlap");
static_assert((EXPLICIT_BIT_MASK | FRACTION_MASK) == SIG_MASK, "the explicit bit and the fractional part should cover the whole significand");

```

File: libc/src/__support/FPUtil/x86_64/LongDoubleBits.h [hunk 3]
```diff
@@ -37,20 +37,21 @@ struct FPBits<long double> : public internal::FPRep<FPType::X86_Binary80> {
...
+  static_assert((EXPLICIT_BIT_MASK | FRACTION_MASK) == SIG_MASK, "mask cover");
   static constexpr StorageType MIN_SUBNORMAL = StorageType(1);
-  // Subnormal numbers include the implicit bit in x86 long double formats.
```
Comment (legrosbuffle) [inline libc/src/__support/FPUtil/x86_64/LongDoubleBits.h:52]: Let's keep the comment.

File: libc/src/__support/FPUtil/FPBits.h [hunk 4]
```diff
@@ -136,6 +136,11 @@ struct FPRepBase : public internal::FPLayout<fp_type> {
   LIBC_INLINE static constexpr StorageType bit_at(int position) {
     return StorageType(1) << position;
   }
+  LIBC_INLINE static constexpr StorageType merge(StorageType a, StorageType b,
```
Comment (legrosbuffle) [inline libc/src/__support/FPUtil/FPBits.h:139]: Please add a comment, it's not obvious what the naming means, and I don't think it's standard terminology.

**Review by legrosbuffle — APPROVED** [review]

**Pull request merged**